src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala

```diff
@@ -528,6 +534,161 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
+                  // thus increasing conflict probability
+                  DeltaSQLConf.MERGE_REPARTITION_BEFORE_WRITE.key -> "true") {
+        closure(dir.getAbsolutePath)
+      }
+    }
+  }
+
+  val concurrentCompactionTestPartitions = 3
+  val concurrentCompactionTestRangeSize = 10
+  val concurrentCompactionTestNumIterations = 10
+
+  def createConcurrentCompactionTestTable(location: String): Unit = {
+    spark.range(0)
+            .withColumn("par", $"id" % concurrentCompactionTestPartitions)
+            .coalesce(1)
+            .write.partitionBy("par").format("delta").save(location)
+  }
+
+  def executeConcurrentCompactionTestInsertOnlyMerge(location: String, i: Int): Unit = {
+    val df = spark.range(i, i + concurrentCompactionTestRangeSize)
+            .withColumn("par", $"id" % concurrentCompactionTestPartitions)
+    val table = io.delta.tables.DeltaTable.forPath(spark, location)
+    table.alias("t").merge(df.alias("s"), "slowMergeCondition(s.par=t.par AND s.id=t.id)")
+            .whenNotMatched().insertAll()
+            .execute()
+  }
+
+  def executeConcurrentCompactionTestUpsertMerge(location: String, i: Int): Unit = {
+    val df = spark.range(i, i + concurrentCompactionTestRangeSize)
+            .withColumn("par", $"id" % concurrentCompactionTestPartitions)
+    val table = io.delta.tables.DeltaTable.forPath(spark, location)
+    table.alias("t").merge(df.alias("s"), "slowMergeCondition(s.par=t.par AND s.id=t.id)")
+            .whenNotMatched().insertAll()
+            .whenMatched().updateAll()
+            .execute()
+  }
+
+  def assertConcurrentCompactionTestResults(location: String): Unit = {
+    val expectedDF = spark
+            .range(concurrentCompactionTestNumIterations + concurrentCompactionTestRangeSize - 1)
+            .withColumn("par", $"id" % concurrentCompactionTestPartitions)
+            .orderBy("id")
+    val deltaResultDF = spark.read.format("delta").load(location).orderBy("id")
+    checkAnswer(deltaResultDF, expectedDF)
+    // make sure no extra files were left after last vacuum
+    // this is a safety check to see that we are not leaving any untracked files in the table
+    val parquetResultDF = spark.read.parquet(location).orderBy("id")
+    checkAnswer(parquetResultDF, expectedDF)
+  }
+
+  def runCompaction(location: String): Unit = {
+    val table = io.delta.tables.DeltaTable.forPath(spark, location)
+    val lastTableOperation = table.history(1).select("operation").head().getString(0)
+    if (lastTableOperation == "MERGE") {
+      spark.read.format("delta")
+              .load(location)
+              .coalesce(1)
+              .write.format("delta")
+              .mode("overwrite")
+              .option("dataChange", "false")
+              .save(location)
+    }
+  }
+
+  class NoException extends Throwable
+
+  def runConcurrentCompactionTest[ExpectedExceptionType: ClassTag]
+  (location: String, executeMergeTest: (String, Int) => Unit): Unit = {
```

Comment:
nit: input argument formatting (it's hard to read it with all these Scala goodies :))

Reply:
How about this?
```scala
  def runConcurrentCompactionTest[ExpectedExceptionType: ClassTag](
      location: String,
      executeMergeTest: (String, Int) => Unit
  ): Unit = {
```

Or do you prefer this?
```scala
  def runConcurrentCompactionTest[ExpectedExceptionType: ClassTag](
      location: String,
      executeMergeTest: (String, Int) => Unit): Unit = {
```

Reply:
My preference would be the latter.

```
def runConcurrentCompactionTest[ExpectedExceptionType: ClassTag](
      location: String,
      executeMergeTest: (String, Int) => Unit): Unit = {
```